File: servers/fai/src/fai/utils/jobs.py
import uuid
from collections.abc import (
    Awaitable,
    Callable,
)
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from datetime import (
    UTC,
    datetime,
)
from enum import Enum
from typing import (
    Any,
)

from src.settings import LOGGER


class JobStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class Job:
    id: str
    status: JobStatus
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    result: Any | None = None
    error: str | None = None


class JobManager:
    def __init__(self, max_workers: int = 4):
        self._jobs: dict[str, Job] = {}
        self._executor = ThreadPoolExecutor(max_workers=max_workers)

    def create_job(self) -> str:
        job_id = str(uuid.uuid4())
        job = Job(id=job_id, status=JobStatus.PENDING, created_at=datetime.now(UTC))
        self._jobs[job_id] = job
        LOGGER.info(f"Created job {job_id}")
        return job_id

    async def execute_job(self, job_id: str, func: Callable[..., Awaitable[Any]], *args: Any, **kwargs: Any) -> None:
        if job_id not in self._jobs:
            raise ValueError(f"Job {job_id} not found")

        job = self._jobs[job_id]
        job.status = JobStatus.IN_PROGRESS
        job.started_at = datetime.now(UTC)

        LOGGER.info(f"Starting job {job_id}")

        try:
            result: Any = await func(*args, **kwargs)

            job.status = JobStatus.COMPLETED
            job.completed_at = datetime.now(UTC)
            job.result = result
            LOGGER.info(f"Job {job_id} completed successfully")

        except Exception as e:
            job.status = JobStatus.FAILED
            job.completed_at = datetime.now(UTC)
            job.error = str(e)
            LOGGER.exception(f"Job {job_id} failed: {e}")

    def get_job_status(self, job_id: str) -> Job | None:
        return self._jobs.get(job_id)

    def cleanup_old_jobs(self, max_age_hours: int = 24) -> None:
        cutoff = datetime.now(UTC).timestamp() - (max_age_hours * 3600)
        jobs_to_remove = []

        for job_id, job in self._jobs.items():
            if job.created_at.timestamp() < cutoff:
                jobs_to_remove.append(job_id)

        for job_id in jobs_to_remove:
            del self._jobs[job_id]
            LOGGER.info(f"Cleaned up old job {job_id}")


job_manager = JobManager()

File: packages/fern-docs/bundle/src/jobs.ts
import { NextResponse } from "next/server";

import { kv } from "@vercel/kv";
import { randomUUID } from "crypto";

interface JobStatus {
  id: string;
  status: "pending" | "in_progress" | "completed" | "failed";
  created_at: number;
  started_at?: number;
  completed_at?: number;
  error?: string;
  result?: any;
}

export class JobManager {
  static async createJob(domain: string): Promise<string> {
    const job: JobStatus = {
      id: randomUUID(),
      status: "pending",
      created_at: Date.now(),
    };

    await kv.hset(domain, { [`tpuf_job`]: job });
    return job.id;
  }

  static async executeJob<T>(
    domain: string,
    task: () => Promise<T>
  ): Promise<void> {
    const job = await kv.hget<JobStatus>(domain, `tpuf_job`);
    if (!job) {
      throw new Error("Job not found");
    }

    job.status = "in_progress";
    job.started_at = Date.now();
    await kv.hset(domain, { [`tpuf_job`]: job });

    try {
      const result = await task();
      job.status = "completed";
      job.completed_at = Date.now();
      job.result = result;
    } catch (error) {
      job.status = "failed";
      job.completed_at = Date.now();
      job.error = error instanceof Error ? error.message : String(error);
    }

    await kv.hset(domain, { [`tpuf_job`]: job });
  }

  static async getJobStatus(domain: string): Promise<JobStatus | undefined> {
    const job = await kv.hget<JobStatus>(domain, `tpuf_job`);
    return job ?? undefined;
  }

  static async getJob(domain: string): Promise<JobStatus | undefined> {
    const job = await kv.hget<JobStatus>(domain, `tpuf_job`);
    return job ?? undefined;
  }
}

export function createJobResponse(
  message: string = "Job created successfully",
  job_id: string
) {
  return NextResponse.json({
    message,
    job_id,
  });
}

export function createJobStatusResponse(job: JobStatus | undefined) {
  if (!job) {
    return NextResponse.json({ error: "Job not found" }, { status: 404 });
  }

  return NextResponse.json({
    id: job.id,
    status: job.status,
    completed: job.status === "completed",
    failed: job.status === "failed",
    created_at: job.created_at,
    started_at: job.started_at,
    completed_at: job.completed_at,
    error: job.error,
    result: job.result,
  });
}
